Add builtin len() support of tf.data.Dataset for autograph

yongtang opened the pull request:
While working on autograph was looking for calculating the average
of a tf.data.Dataset, noticed that `len()` in autograph
only support tensor, tensorarray and tensorlist, but not tf.data.Dataset.

I think it makes sense to add tf.data.Dataset as well as `len()` is too widely
used in many cases.

This PR adds `len()` support of tf.data.Dataset for autograph.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add builtin len() support of tf.data.Dataset for autograph

While working on autograph was looking for calculating the average
of a tf.data.Dataset, noticed that `len()` in autograph
only support tensor, tensorarray and tensorlist, but not tf.data.Dataset.

I think it makes sense to add tf.data.Dataset as well as `len()` is too widely
used in many cases.

This PR adds `len()` support of tf.data.Dataset for autograph.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for builtin len() support of tf.data.Dataset in autograph

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

mdanatg commented on tensorflow/python/autograph/operators/py_builtins.py:
```diff
@@ -278,6 +280,10 @@ def raise_zero_rank_error():
 
 
+def _tf_dataset_len(s):
+  return s.reduce(constant_op.constant(0, dtype=dtypes.int32), lambda x, _: x + 1)
```
Please use the cardinality function instead, it can be much more efficient: https://github.com/tensorflow/tensorflow/blob/e5bf8de410005de06a7ff5393fafdf832ef1d4ad/tensorflow/python/data/experimental/ops/cardinality.py#L33
It can also protect against infinite loops (repeated datasets will cycle forever).

If the function returns INFINITE, then we should raise an error.

If it returns UNKNOWN, then we have two options:
 * raise an error outright (safer)
 * fall back to counting with reduce, but in that case we'd need to limit the number of iterations in the eventuality that the dataset might be actually infinite

mdanatg commented:
This is great! Please see the note about efficiency and dealing with infinite datasets.

yongtang committed:
Use cardinality to process the infinite case, as was suggested by review comment

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fallback to reduce to process the unknown cardinality case

The reason for this consideration is that len() of infinite list does exist
so there is a risk with python itself:
```
python3
Python 3.7.3 (default, Dec 13 2019, 19:58:14)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def fib():
...     a, b = 0, 1
...     while True:
...         a, b = b, a + b
...         yield a
...
>>> len(fib())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'generator' has no len()
>>> len(list(fib()))
^CTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in fib
KeyboardInterrupt
>>>

```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add test case for unknown and infinite cadinality of dataset

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Pylint fix

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @mdanatg for the review! I have updated the PR to use cardinality as the default method to find out the length, and throw out error if INFINITE is encountered.

In case of UNKNOWN, the PR falls back to reduce. There is a risk of running into infinite as well. Though even in python's case a len() that causes the infinite is also possible:

```
python3
Python 3.7.3 (default, Dec 13 2019, 19:58:14)
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def fib():
...     a, b = 0, 1
...     while True:
...         a, b = b, a + b
...         yield a
...
>>> len(fib())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: object of type 'generator' has no len()
>>> len(list(fib()))
^CTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 4, in fib
KeyboardInterrupt
>>>

```

Please take a look and let me know if there are any issues.

mdanatg commented on tensorflow/python/autograph/operators/py_builtins.py:
```diff
@@ -278,6 +281,24 @@ def raise_zero_rank_error():
+    return d.reduce(constant_op.constant(0, dtype=dtypes.int32), lambda x, _: x + 1)
+
+  l = control_flow_ops.cond(l != cardinality.INFINITE, lambda: l, raise_infinite_cardinality_error)
+
```
We should be able to make this even more efficient with something like this:

```
with ops.control_dependencies([control_flow_ops.Assert(l != cardinality.INFINITE, [msg])]):
  l = tf.identity(l)
```

That way, we avoid the extra tf.cond altogether.

yongtang replied:
Thanks @mdanatg. Updated.

mdanatg commented on tensorflow/python/autograph/operators/py_builtins_test.py:
```diff
@@ -123,6 +123,28 @@ def test_len(self):
+    dataset = dataset_ops.DatasetV2.range(5).repeat().batch(2)
+    with self.assertRaises(errors_impl.InvalidArgumentError):
+      _ = py_builtins.len_(dataset)
+    with self.cached_session() as sess:
```
I think this latter half of the test is redundant. Perhaps did you want to test it in graph mode? To do that, run it inside a tf.function.

yongtang replied:
Thanks @mdanatg. Done.

mdanatg commented:
Interesting! I think in our case, a dataset with unknown length would be the equivalent of the generator in the Pyhton example. Wrapping an infinite generator into a list is indeed problematic, but I think for datasets such operation doesn't exist yet - it would be something in the lines of a `dataset.stack()`-like operation. So I think it would be totally fine to raise an error for unknown as well.

mdanatg reviewed (commented):
Looks good! A few minor items remaining to clean up.

yongtang committed:
Throw out error for cardinality == UNKNOWN as well

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update PR based on review comment, for using identity and using tf.function for graph mode

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @mdanatg for the review. The PR has been updated with comment addressed. 

I also updated the PR to thrown out error in case of UNKNOWN (together with INFINITE case). Let me know if there are any other issues.

mdanatg reviewed (approved):
Sounds good, thank you. Should it turn out that too many datasets use `filter`, it will be good to remember that the `reduce` option is available. Feel free to record that as a TODO.

mdanatg commented on tensorflow/python/autograph/operators/py_builtins_test.py:
```diff
@@ -123,6 +123,43 @@ def test_len(self):
       tl = py_builtins.len_(data_structures.tf_tensor_list_new([3, 4, 5]))
       self.assertEqual(self.evaluate(tl), 3)
 
+  def test_len_dataset(self):
```
Sorry, I missed an issue that is blocking internal tests - these can also run with eager mode off. To fix that, could you modify the checks as follows:

`self.assertEqual(self.evaluate(py_builtins.len_(dataset)), 3)`
`self.assertEqual(self.evaluate(test_fn()), 3)`

and so on

That way they'll be treated consistently in both eager and graph modes.

yongtang replied:
@mdanatg Thanks. The PR has been updated.

yongtang committed:
Update test with self.evaluate(), as was suggested in review

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @mdanatg for the review, please take a look and let me know if there are any other issues.
